Enable C++20

yol opened the pull request:
## Description
Enable C++20 and fix all resulting compile problems

Original work by @lrusak 

## Motivation and context
We are looking into using the new date functionality of C++20 and this may allow us not to have to use a 3rdparty dependency.

## How has this been tested?
Compile+run on Linux

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

lrusak committed:
[cmake] use c++20

lrusak committed:
c++20: remove volatile usage

In CacheStrategy.h, it is unclear why volatile was ever used.
In AEUtil.h, the std::atomic will be safer and likely more correct
than the previous usage of volatile.

lrusak committed:
c++20: need to use [=, this] to pass reference for lambda

lrusak committed:
c++20: u8 literal has different meaning

lrusak committed:
StringUtils: Mark passed format strings as runtime, fix type conversion

lrusak committed:
SMBWSDiscoveryListener: use std::string_view for constexpr fmt::format usage

yol committed:
[depends] use c++20

yol committed:
Fix wrong format string with unescaped {}

wsnipex commented:
note that we already have NDK 25 for android available, which supports c++20. We'd just have to enable it by default.
Test build: https://jenkins.kodi.tv/job/android-arm64-docker/5733/console

wsnipex commented on tools/depends/configure.ac:
```diff
@@ -667,9 +667,9 @@ fi
+AX_CXX_COMPILE_STDCXX([20], [noext], [mandatory])
 CXX="$CXX_CACHED"
-platform_cxxflags="${platform_cxxflags} -std=c++17"
+platform_cxxflags="${platform_cxxflags} -std=c++20"
```
we could keep 17 for depends for now, should get rid of the build issues

sundermann replied:
This breaks `smctemp` on macOS. We override `CXXFLAGS` inside our makefile [here](https://github.com/xbmc/xbmc/blob/master/tools/depends/target/smctemp/Makefile#L19) but the native makefile needs `-std=c++17` to be [present](https://github.com/narugit/smctemp/blob/0.2.1/Makefile#L2C13-L2C49). I guess we need to override `CXXFLAGS` because of the different architectures but we should probably have `-Wall -std=c++17 -g -framework IOKit` as default `CXXFLAGS` in our makefile to have the same flags as upstream.

yol replied:
thanks for finding this, let's see what happens

yol commented:
I really like how WebOS was the only platform to build on the first try :sunglasses:

yol commented:
Can someone with MacOS knowledge say if we can update XCode? TVOS (at least) seems to be failing due to some compiler bug that's probably fixed already. @fuzzard @phunkyfish?

yol committed:
[windows] Add /utf-8 flag for MSVC

This fixes C4566 warnings on Windows/MSVC and also corrects unit test failures
that would start showing up with C++20.

While the previous behavior might have worked sometimes by accident, this now
sets proper encodings for source and execution character sets.

Related reading:
* https://pspdfkit.com/blog/2021/string-literals-character-encodings-and-multiplatform-cpp/
* https://devblogs.microsoft.com/cppblog/new-options-for-managing-character-sets-in-the-microsoft-cc-compiler/
* https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4566?view=msvc-170

yol committed:
c++20: Fix deprecated call to shared_ptr::unique

yol committed:
[windows] c++20: Do not compile C files as C++

C and C++ aren't actually compatible and this breaks the build starting with C++20 (due to cc_overlay.c using u8 literals).
When these files are now compiled as C files, they cannot use the C++ precompiled headers, so just disable PCH for those few files.

Commit:
[buildsteps] Bump ios/tvos to Xcode 15.2 and SDK 17.2

wsnipex commented:
> > note that we already have NDK 25 for android available, which supports c++20. We'd just have to enable it by default. Test build: https://jenkins.kodi.tv/job/android-arm64-docker/5733/console
> 
> I have zero idea about Android, can you do that? Or could that lead to any problems?

I've enabled it for the PR builder for now. Let's see if it works.
note: param NDK_VERSION=25.1.8937393 in the BuildMulti-PR jenkins config for the android builders

wsnipex commented:
jenkins build this please

Commit:
[tools/depends][toolchain] propagate minver for ios/tvos to toolchain from configure

Commit:
[buildsteps] Bump macos to Xcode 15.2 and SDK 14.2

fuzzard commented:
All android jobs using ndk 25 will need the additional swab commit from https://github.com/xbmc/xbmc/pull/22116 to function

lrusak commented:
Should we do the NDK/SDK bumps for various platforms in separate precursor PR's?

phunkyfish commented:
@fuzzard do you remember where we make the changes for binary addons?

sundermann commented:
@phunkyfish It should be set from [`Toolchain_binaddons.cmake`](https://github.com/xbmc/xbmc/blob/master/tools/depends/target/Toolchain_binaddons.cmake.in#L84) which uses `platform_cxxflags`. I guess we do need to set `set(CMAKE_CXX_STANDARD 20)` now to force c++20 in binary addons.

neo1973 commented:
Building with C++20 causes this crash (also crashes in non sanitized build):

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==102770==ERROR: AddressSanitizer: SEGV on unknown address 0x5a73e6eed620 (pc 0x7a0a0515c374 bp 0x7ffcf5dd76d0 sp 0x7ffcf5dd7688 T0)
==102770==The signal is caused by a WRITE memory access.
    #0 0x7a0a0515c374 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_Alloc_hider::_Alloc_hider(char*, std::allocator<char> const&) /usr/src/debug/gcc/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.h:189:25
    #1 0x5a73e5ce1971 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::basic_string<std::allocator<char>>(char const*, std::allocator<char> const&) /usr/include/c++/13.2.1/bits/basic_string.h:632:9
    #2 0x7a0a043e54e6 in __static_initialization_and_destruction_0 /usr/src/debug/p8-platform/platform-p8-platform-2.1.0.1/src/util/StringUtils.cpp:51:40
    #3 0x7a0a043e54e6 in _GLOBAL__sub_I_StringUtils.cpp /usr/src/debug/p8-platform/platform-p8-platform-2.1.0.1/src/util/StringUtils.cpp:1149:1
    #4 0x7a0a07c921bd  (/lib64/ld-linux-x86-64.so.2+0x51bd) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
    #5 0x7a0a07c922ab  (/lib64/ld-linux-x86-64.so.2+0x52ab) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
    #6 0x7a0a07ca8bbf  (/lib64/ld-linux-x86-64.so.2+0x1bbbf) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
```
The problem is that both Kodi and p8-platform define `StringUtils::Empty` but because of the different C++ standards they are not compatible:

https://github.com/Pulse-Eight/platform/blob/5a786dd580972a2dc89941ea60dc08c0314e0e4d/src/util/StringUtils.cpp#L51

https://github.com/xbmc/xbmc/blob/e70f0d83f3ae5e5026cf6b6402c8d7e05662bc8a/xbmc/utils/StringUtils.cpp#L77

The good thing is that p8-platform doesn't seem to use `StringUtils::Empty` at all so I'll make a PR to remove the symbol.

phunkyfish commented:
> @phunkyfish It should be set from [`Toolchain_binaddons.cmake`](https://github.com/xbmc/xbmc/blob/master/tools/depends/target/Toolchain_binaddons.cmake.in#L84) which uses `platform_cxxflags`. I guess we do need to set `set(CMAKE_CXX_STANDARD 20)` now to force c++20 in binary addons.

We shouldn’t do this until the binary addons are branch for P, we need a name first 😉

phunkyfish commented:
> Building with C++20 causes this crash (also crashes in non sanitized build):
> 
> ```
> AddressSanitizer:DEADLYSIGNAL
> =================================================================
> ==102770==ERROR: AddressSanitizer: SEGV on unknown address 0x5a73e6eed620 (pc 0x7a0a0515c374 bp 0x7ffcf5dd76d0 sp 0x7ffcf5dd7688 T0)
> ==102770==The signal is caused by a WRITE memory access.
>     #0 0x7a0a0515c374 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::_Alloc_hider::_Alloc_hider(char*, std::allocator<char> const&) /usr/src/debug/gcc/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.h:189:25
>     #1 0x5a73e5ce1971 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::basic_string<std::allocator<char>>(char const*, std::allocator<char> const&) /usr/include/c++/13.2.1/bits/basic_string.h:632:9
>     #2 0x7a0a043e54e6 in __static_initialization_and_destruction_0 /usr/src/debug/p8-platform/platform-p8-platform-2.1.0.1/src/util/StringUtils.cpp:51:40
>     #3 0x7a0a043e54e6 in _GLOBAL__sub_I_StringUtils.cpp /usr/src/debug/p8-platform/platform-p8-platform-2.1.0.1/src/util/StringUtils.cpp:1149:1
>     #4 0x7a0a07c921bd  (/lib64/ld-linux-x86-64.so.2+0x51bd) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
>     #5 0x7a0a07c922ab  (/lib64/ld-linux-x86-64.so.2+0x52ab) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
>     #6 0x7a0a07ca8bbf  (/lib64/ld-linux-x86-64.so.2+0x1bbbf) (BuildId: 9e3efada0c5a9fde834899c54ccdb8b2fce628e4)
> ```
> 
> The problem is that both Kodi and p8-platform define `StringUtils::Empty` but because of the different C++ standards they are not compatible:
> 
> https://github.com/Pulse-Eight/platform/blob/5a786dd580972a2dc89941ea60dc08c0314e0e4d/src/util/StringUtils.cpp#L51
> 
> https://github.com/xbmc/xbmc/blob/e70f0d83f3ae5e5026cf6b6402c8d7e05662bc8a/xbmc/utils/StringUtils.cpp#L77
> 
> The good thing is that p8-platform doesn't seem to use `StringUtils::Empty` at all so I'll make a PR to remove the symbol.
> 
> Edit: Maybe it's not so easy as `StringUtils::Empty` could be intentionally part of the public interface? Need to dig deeper...

There is also a version of StringUtils in the binary addons API, we should check that too.

phunkyfish commented:
> > > @phunkyfish It should be set from [`Toolchain_binaddons.cmake`](https://github.com/xbmc/xbmc/blob/master/tools/depends/target/Toolchain_binaddons.cmake.in#L84) which uses `platform_cxxflags`. I guess we do need to set `set(CMAKE_CXX_STANDARD 20)` now to force c++20 in binary addons.
> > 
> > 
> > Once all the platforms build we can do this. Then we can build with addons. But we won’t be able to merge this PR until the binary addons are branched for P. Nor can we make changes to said addons until then.
> 
> Alright, in that case we still have to add `set(CMAKE_CXX_STANDARD 17)` to keep things as is since `platform_cxxflags` no longer contains `-std=c++17`

This is the problem. We can’t run binary addons on a different version of c++. They need to be built with the same version. But that’s step 2. We’re just on step 1 right now 😉

yol committed:
[depends] set default C++ standard for binaddons build

yol commented:
* FreeBSD build is OK after upgrade to FreeBSD 14.0
* patches by @fuzzard picked for macOS/iOS
* CMAKE_CXX_STANDARD 17 added to binaddons
* Windows is building now but failing a charset conversion test
* Android still failing due to NDK - can we merge #22116 ? will that fix the issues so that we can update NDK?

> This is the problem. We can’t run binary addons on a different version of c++. They need to be built with the same version. 

Why is that? I don't see why you shouldn't be able to link objects built with `-std=c++17` and `-std=c++20` together. In fact, I think this is common practice given that most libraries will not build with `-std=c++20`  themselves yet, but you can still use them in your C++20 programs. 

> The problem is that both Kodi and p8-platform define StringUtils::Empty but because of the different C++ standards they are not compatible:

I'm a bit fuzzy here, how does that work? How do they both define the same symbol in the first place?

yol commented:
jenkins build this please

neo1973 commented:
Regarding the crash, it only seems to happen when I build Kodi with Clang. With GCC it's fine. I wonder what the problem is, both are building with libstdc++.

yol commented:
clang seems to default to c++98, interestingly. will try to pull https://github.com/Pulse-Eight/platform/pull/46 into depends as patch

yol committed:
[depends] Incorporate p8-platform PR #46

fuzzard commented on cmake/modules/FindP8Platform.cmake:
```diff
@@ -14,7 +14,8 @@ if(NOT P8Platform::P8Platform OR P8Platform_FIND_REQUIRED)
                 "${CORE_SOURCE_DIR}/tools/depends/target/${MODULE_LC}/002-all-fixcmakeinstall.patch"
-                "${CORE_SOURCE_DIR}/tools/depends/target/${MODULE_LC}/003-all-cmake_tweakversion.patch")
+                "${CORE_SOURCE_DIR}/tools/depends/target/${MODULE_LC}/003-all-cmake_tweakversion.patch"
+                "${CORE_SOURCE_DIR}/tools/depends/target/${MODULE_LC}/003-all-fix-cxx-standard.patch")
```
004-all-fix-cxx-standard.patch

yol replied:
ffs :joy:

phunkyfish commented:
Looks like StrungUtils fails on OSX:

```
07:09:23 /Users/Shared/jenkins/workspace/OSX-64/build/build/build-p8-platform/src/build-p8-platform/src/util/StringUtils.cpp:469:49: error: expected expression
07:09:23   str.erase(::find_if(str.rbegin(), str.rend(), [](char s) { return isspace_c(s) == 0; }).base(), str.end());
```

what is the change in behaviour in C++20?

yol commented:
> what is the change in behaviour in C++20?

this is not c++20-related at all, I think it happens because I removed `-std=c++17` from the depends cxxflags - so clang++ defaults to c++98, which does not have lambdas.

yol commented:
jenkins build this please

fuzzard commented on cmake/scripts/windowsstore/ArchSetup.cmake:
```diff
@@ -68,6 +68,9 @@ endif()
 # -------- Compiler options ---------
 
+# Allow to use UTF-8 strings in the source code, disable MSVC charset conversion
+add_options(CXX ALLBUILDS "/utf-8")
```
Presumably this is meant to be ALL_BUILDS?

joseluismarti commented:
As suggested by @wsnipex  set param `NDK_VERSION=25.2.9519653` in jenkins config for the android builders

yol commented:
It seems NDK 25 is not recent enough. @wsnipex is it possible to get NDK 26?

yol commented:
jenkins build this please

joseluismarti commented:
Jenkins builders are using NDK 21, see console output:
 ```
--with-ndk-path=/home/jenkins/android-tools/android-sdk-linux/ndk/21.4.7075529 
```

yol commented:
@joseluismarti yes and no, I did a separate build with NDK 25 and that still fails -> https://jenkins.kodi.tv/job/android-arm-docker/5988/

On another note, this PR now builds OK and all platforms but Android :partying_face: 

@neo1973 can you check if your runtime issue is now solved as well?

joseluismarti commented:
I've managed to build locally with NDK 25 (LLVM 14), with the inconvenience that `contexpr` has to be removed from these functions:
https://github.com/xbmc/xbmc/blob/1ed0a37ecde730d5ce4bc11e71cd9188cc5b1c29/xbmc/utils/StringUtils.h#L79
https://github.com/xbmc/xbmc/blob/1ed0a37ecde730d5ce4bc11e71cd9188cc5b1c29/xbmc/utils/StringUtils.h#L85
and da0fd26cbf2335064b9ced0ec96601041a4ae888 has to be added.

Allows to identify the level of compatibility with C++20, I don't know if this test is useful for you :+1:.

neo1973 commented:
> can you check if your runtime issue is now solved as well?

Unfortunately it's still present :cry:

joseluismarti commented:
My test build seems to work fine on Android 12

yol commented:
> I've managed to build locally with NDK 25 (LLVM 14), with the inconvenience that contexpr has to be removed from these functions

Yeah AFAICT this is due to libc++ not being "compliant" enough yet in NDK 25. If they can't even support that level of usage, it's unlikely we'll get much C++20 benefit from that.

> Allows to identify the level of compatibility with C++20, I don't know if this test is useful for you 👍.

It is, thanks!

> Unfortunately it's still present 😢

OK I will build with clang and see what happens...

yol commented:
@neo1973 is there something specific I need to do to trigger the bug?

neo1973 commented:
I believe key are:

- building with Clang
- linking against libcec which in turn links p8-platform

neo1973 commented:
I think I found the significant difference. When I build with Clang in C++17 or with GCC the `StringUtils::Empty` symbol is in the .bss ELF section. With Clang in C++20 its in the .data.rel.ro section. If I change the code to force the symbol into the .data.rel.ro section for GCC then I see the exact same crash as well.

I'm not an expert on that matter but normally it shouldn't matter in which ELF section the compiler (or linker?) moves the symbol. But I believe having the same symbol in Kodi and a shared library is undefined behavior, the ELF section makes a subtile difference and it's just luck that it hasn't bitten us so far.

Don't take this at face value, as said I'm not an expert :slightly_smiling_face:

neo1973 commented:
I came across https://github.com/llvm/llvm-project/issues/72361 which isn't the same problem but similar and helps to understand our problem. What's happening for us is that `StringUtils::Empty` has been moved into a read-only data section in the Kodi binary. When `p8-platform` is loaded it wants to initialize `StringUtils::Empty` because theirs is in the .bss section. But because a symbol can only exist once it tries to initialize the already loaded symbol in read-only memory and that causes a segfault.

sundermann commented:
Does compiling `p8-platform` in C++20 as well fix it?

neo1973 commented:
It does. When building it as C++17 the `StringUtils::Empty` symbol is in the .bss section of libp8-platform.so.2.1.0 and in C++20 mode in the .data section. It's the same for GCC and Clang.

yol commented:
I see two options re. p8-platform:

1. Do a workaround like https://android-review.googlesource.com/c/platform/system/unwinding/+/3010579/2/libunwindstack/include/unwindstack/SharedString.h
2. Wrap "our" kodi StringUtils in a `KODI::UTILS` namespace and then import that namespace to the global one. That should make sure the symbols end up in `KODI::UTILS` in the symbol table, but we don't have to touch every line of code that uses StringUtils today. Not sure if this would work, but might make sense to try.

I still can't reproduce it though :/ My clang puts StringUtils::Empty into `.data`. Any ideas why? What clang version are you using @neo1973? I'm on clang 16.0.6 with gcc-libs 13.2.1 / libstdc++ 6.0.32.

neo1973 commented:
> What clang version are you using @neo1973?

Clang 17.0.6 with gcc-libs 13.2.1 / libstdc++ 6.0.32.

fuzzard commented:
Jenkins build this please

yol commented:
@neo1973 then I guess the problem starts with clang 17, which I can't test at the moment. I've pushed a commit that might fix it. can you try?

neo1973 commented:
> then I guess the problem starts with clang 17, which I can't test at the moment. I've pushed a commit that might fix it. can you try?

Tried and doesn't crash anymore :+1:

fuzzard commented:
Ship it!

yol commented:
unfortunately I have to clean up the messy commits now before shipping it :see_no_evil:

yol committed:
libUPnP: fix template definitions for c++20

yol committed:
libUPnP: Remove deprecated volatile usage

With the atomic_int, the mutex is no longer necessary as well.

yol committed:
libUPnP: Fix deprecated generation of copy assignment operator

yol committed:
libUPnP: Fix warning about double promotion for NPT_TimeStamp

yol committed:
CLog: allow using constexpr formatter and use fmt::vformat

yol commented:
This should be fairly clean and complete now; I only have to add the clang crash fix that I seem to have lost during rebase :D I think I still have it on my other PC.

yol commented:
@fuzzard can you look into the iOS failures? 
13:46:54 /Users/Shared/jenkins/workspace/IOS-ARM64/xbmc/platform/darwin/ios/LaunchScreen.storyboard: error: iOS 17.2 Platform Not Installed.

fuzzard commented:
Jenkins build this please

fuzzard commented:
@yol good to go for those two jobs now

yol commented:
@fuzzard Thanks!

So what do we want to do with p8platform? I personally think that it's a terrible idea that it declares StringUtils without namespace and it has come to bite us now. However, if we are going to change that, it requires a major p8platform release (moving StringUtils to the `P8PLATFORM` namespace) along with a minor libcec release (because otherwise it wouldn't build with newer p8platform). This is quite a bit of hassle, but IMO a bit more future-proof than working around the problem in Kodi.

@opdenkamp @malard thoughts?

yol commented:
@opdenkamp AFAICT libCEC uses StringUtils from p8platform, so when you change the namespace there, it's not going to find the symbols any more

lrusak commented:
Could we not try and remove the use of p8-platform completely? How many packages/add-ons still use it?

yol commented:
It's a dependency of libcec

yol commented:
Yep, but as long as you have a libCEC (or other thing) that still uses the old library, we will have the Kodi crash at startup, so for the actual problem to go away, we'd need to bump libCEC as well.

opdenkamp commented:
I haven't been reading this whole PR, sorry, but what I saw was that both define `StringUtils::Empty` but with different values. If you move whatever Kodi needs into its own namespace, so it becomes `P8PLATFORM::StringUtils::Empty`, and libCEC keeps using what it's using now, won't that fix the problem.

I don't want to rush a major release of libCEC at this time without checking all the open PRs/reports first, and that will take some time.

yol commented:
>  but what I saw was that both define StringUtils::Empty but with different values

Not exactly; the value is the same but one is compiled with C++17 and one with C++20 leading to a linking problem with the newest clang version. In any case, what we are doing is a violation of ODR and thus undefined behavior.

> If you move whatever Kodi needs into its own namespace, so it becomes P8PLATFORM::StringUtils::Empty, and libCEC keeps using what it's using now, won't that fix the problem.

Yes, we could also move the StringUtils in Kodi, so that they become `KODI::StringUtils`. Kodi itself does not directly need p8platform (AFAIK). I tried briefly to move the class but it lead to a lot of changes so I gave up. All things considered, this might be the easier way still. If possible, I still think that long-term p8platform not exporting anything in the global C++ namespace would be a good idea.

yol committed:
Move ColorUtils, FontUtils, Moving Speed to KODI::UTILS

Usage of namespaces is split for utilities is split between UTILS
and KODI::UTILS at the moment. The aim here is to settle on
one variant (KODI::UTILS), as otherwise lookup conflicts will
occur when both namespaces are declared.

yol committed:
Move StringUtils to KODI::UTILS namespace

yol commented:
I decided to go the route of moving StringUtils to KODI::UTILS and then importing it into the global namespace. This still needed some other changes as a consequence due to namespace clashes with utils declared in UTILS (as opposed to KODI::UTILS), but overall I think the changeset is a lot smaller than if we qualify StringUtils everywhere. Let me know what you all think.

fuzzard reviewed (approved):
Had a bit of a glance through commits. As best i can tell it all seems sane.

I hate the fact we have libupnp (Neptune and Platinum) in tree and we are patching the shit out of it, but thats not going to change anytime soon, so it is what it is.

Best way to test it is to ship it into nightlies, so lets go. Might want to keep on eye on binary addons post merge, but im sure it'll be ok...

yol commented:
@lrusak good to merge?

sundermann commented on tools/depends/target/p8-platform/Makefile:
```diff
@@ -2,7 +2,8 @@ include ../../Makefile.include P8-PLATFORM-VERSION ../../download-files.include
                               002-all-fixcmakeinstall.patch \
-                              003-all-cmake_tweakversion.patch
+                              003-all-cmake_tweakversion.patch \
+							  004-all-fix-cxx-standard.patch
```
Spacing is off here

sundermann commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -1947,3 +1950,5 @@ std::string StringUtils::CreateFromCString(const char* cstr)
   return cstr != nullptr ? std::string(cstr) : std::string();
 }
+
+} // namespace KODI::UTILS
```
No newline at end of file

sundermann reviewed (approved):
Changes look reasonable to me.

Runtime tested on webOS and MacOS.

phunkyfish commented:
Has anyone tested this with addons on c++17 and kodi on c++20?

yol commented:
AFAICT, after applying this, binary addons are still built with c++17, right?

phunkyfish commented:
> AFAICT, after applying this, binary addons are still built with c++17, right?

I believe so, until we flip the config to build addons with C++20. @AlwinEsch might have some insights into whether or not this could cause problems.

yol commented:
last call, will merge tomorrow.

phunkyfish commented:
Jenkins build with addons please

phunkyfish commented:
jenkins build this with addons please

phunkyfish commented:
> jenkins build this with addons please



> jenkins build this with addons please

@yol, found it!

yol commented:
seemed to be OK, so merging :)

CastagnaIT commented:
i noticed that now on inputstream adaptive repo
CI linux test builds (Debian package test) seem to works with C++20 and so CI linux build is broken due to dependecies CPP20 incompatibility (https://github.com/xbmc/inputstream.adaptive/actions/runs/9542939073/job/26298621948)

im upgrading ISAdaptive addon dependencies to allow build it correctly on C++20 and with the opened PR Debian CI is happy

but i dont know if i have to wait or if i can merge a bump PR, because on this dependency bump:
https://github.com/xbmc/inputstream.adaptive/pull/1582/files
i have specified the DRAPIDJSON_BUILD_CXX20 flag
but kodi addons build system is not full yet enabled for c++20? is in a kind of mixage? so im a bit confused on how to proceeds and if i have to wait or whatelse
